elasticsearch/shard_migration.py

```diff
@@ -0,0 +1,128 @@
+"""
+
+HELP_EXAMPLE = "Example: ./indices_migration.py --host http://localhost:9200"
+
+
+def parse_args():
+    parser = ArgumentParser(
+        prog="Elasticsearch shard migration tool",
+        description=HELP_DESCRIPTION,
+        epilog=HELP_EXAMPLE,
+    )
+    parser.add_argument(
+        "-H",
+        "--host",
+        dest="es_host",
+        default="http://localhost:9200",
+        help="Elasticsearch host.",
+    )
+
+    return parser.parse_args()
+
+
+class Node:
+    def __init__(self, ip, name):
+        self.ip = ip
+        self.name = name
+
+    def __str__(self):
+        return f"{self.name} ({self.ip})"
+
+
+class Elasticsearch:
+    def __init__(self, host):
+        """
+        Initialize Elasticsearch client with the specified host.
+        :param host: Elasticsearch host URL.
+        """
+        self.host = host
+        self.nodes = self.__get_nodes()
+
+    def __get_nodes(self):
+        """
+        Get all nodes in the ES cluster and return a list of Node objects.
+        """
+        url = f"{self.host}/_cat/nodes?format=json"
+        try:
+            LOG.info("Fetching all nodes in the ES cluster.")
+            response = requests.request("GET", url)
+            response.raise_for_status()
+            return [Node(node["ip"], node["name"]) for node in response.json()]
+        except requests.exceptions.RequestException as e:
+            LOG.error(f"Error fetching nodes from ES cluster: {e}")
+            return []
+
+    def migrate_shards(self, nodes):
+        """
+        Exclude selected nodes from new shard allocations in the cluster
+        and migrate all of the existing shards on them to other nodes.
+        :param nodes: List of Node objects to be excluded.
+        """
+        try:
+            ip_list = ",".join([node.ip for node in nodes])
+            LOG.info(f"Migrating shards from nodes: {ip_list}")
+
+            body = {"persistent": {"cluster.routing.allocation.exclude._ip": ip_list}}
+
+            url = f"{self.host}/_cluster/settings"
+            response = requests.request(
+                "PUT",
+                url,
+                headers={"Content-Type": "application/json"},
+                data=json.dumps(body),
+            )
+            response.raise_for_status()
+            LOG.info(f"Successfully started shard migration from nodes: {ip_list}")
+        except requests.exceptions.RequestException as e:
+            LOG.error(f"Error starting shard migration: {e}")
+
+
+def main():
+    args = parse_args()
+
+    es_host = args.es_host
+    es = Elasticsearch(es_host)
+
+    if not es.nodes:
+        LOG.warning("No nodes available in the cluster.")
+        return
+
+    print("\nAvailable Nodes in the Cluster:")
+    for idx, node in enumerate(es.nodes, start=1):
+        print(f"{idx}. {node}")
+
+    selection = input(
+        "\nEnter the numbers of the nodes to exclude, separated by spaces: "
+    )
```

Comment:
Not a fan of that, CLI flags are almost always better. Especially since you can more easily save examples of usage.

Reply:
Also numbers sounds more error prone than giving whole node names.

Reply:
My reasoning was to display first the members of the cluster so the user knows which are they, but now when I think about it that could be also a flag of the CLI.

Reply:
You could just make the default behavior without arguments to print a list.